qa/tasks/cephfs/test_mirroring.py

```diff
@@ -225,11 +225,14 @@ def check_peer_status_renamed_snap(self, fs_name, fs_id, peer_spec, dir_name,
     def check_peer_snap_in_progress(self, fs_name, fs_id,
                                     peer_spec, dir_name, snap_name):
         peer_uuid = self.get_peer_uuid(peer_spec)
-        res = self.mirror_daemon_command(f'peer status for fs: {fs_name}',
-                                         'fs', 'mirror', 'peer', 'status',
-                                         f'{fs_name}@{fs_id}', peer_uuid)
-        self.assertTrue('syncing' == res[dir_name]['state'])
-        self.assertTrue(res[dir_name]['current_sycning_snap']['name'] == snap_name)
+        with safe_while(sleep=1, tries=30, action=f'wait for status') as proceed:
+            while proceed():
+                res = self.mirror_daemon_command(f'peer status for fs: {fs_name}',
+                                                 'fs', 'mirror', 'peer', 'status',
+                                                 f'{fs_name}@{fs_id}', peer_uuid)
+                if('syncing' == res[dir_name]['state'] and \
```

Comment:
You mentioned in https://tracker.ceph.com/issues/64711#note-4 that sync failed for /d0. Have you checked the reason for -EBADFD?

Reply:
Will check that.

Reply:
As discussed, check the client log to see why its hitting an EBADFD.

Reply:
@joscollin ping?

Reply:
I've checked it once, but couldn't find much clue from the PeerReplayer side of the issue back then. Will check again.

Reply:
Set `self.config_set('cephfs-mirror', 'debug client', 20)` in the test_mirroring.py?

Reply:
@vshankar I tried to reproduce this failure by checking out the exact same sha of `/a/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/7580933` and then added `debug client = 20` commit and then rerun the `fs:mirror` suite. Then the test passes: https://pulpito.ceph.com/jcollin-2024-03-22_09:24:18-fs:mirror-wip-jcollin-testing-22032024-distro-default-smithi/. I was expecting a failure with more debug info, but the test passed.

Reply:
@vshankar I've also checked, but it doesn't seem to be related to https://github.com/ceph/ceph/pull/55457.

Reply:
> @vshankar I tried to reproduce this failure by checking out the exact same sha of `/a/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/7580933` and then added `debug client = 20` commit and then rerun the `fs:mirror` suite. Then the test passes: https://pulpito.ceph.com/jcollin-2024-03-22_09:24:18-fs:mirror-wip-jcollin-testing-22032024-distro-default-smithi/. I was expecting a failure with more debug info, but the test passed.

Try running in a batch on 10 (or even 20). If the issue is related to a race, it _might not_ show up in a single run.

Reply:
@vshankar I got the failures reproduced now with `debug client: 20`. 
https://pulpito.ceph.com/jcollin-2024-03-25_04:53:04-fs:mirror-wip-jcollin-testing-22032024-distro-default-smithi/

I've grepped through the logs and they all have valid `(fh)`. Checking the client logs again for this https://tracker.ceph.com/issues/64711#note-4.

Reply:
Here we got:
```
 42498 2024-03-25T05:29:35.795 INFO:teuthology.orchestra.run.smithi155.stderr:dumped fsmap epoch 104
 42499 2024-03-25T05:29:35.861 INFO:teuthology.orchestra.run:Running command with timeout 30
 42500 2024-03-25T05:29:35.862 DEBUG:teuthology.orchestra.run.smithi155:peer status for fs: cephfs> ceph --admin-daemon /var/run/ceph/cephfs-mirror.asok fs mirror peer status cephfs@14 03077e74-e57d-47b6-be66-a4010927dcc5
 42501 2024-03-25T05:29:36.009 INFO:teuthology.orchestra.run.smithi155.stdout:{
 42502 2024-03-25T05:29:36.009 INFO:teuthology.orchestra.run.smithi155.stdout:    "/d0": {
 42503 2024-03-25T05:29:36.009 INFO:teuthology.orchestra.run.smithi155.stdout:        "state": "syncing",
 42504 2024-03-25T05:29:36.009 INFO:teuthology.orchestra.run.smithi155.stdout:        "current_sycning_snap": {
 42505 2024-03-25T05:29:36.009 INFO:teuthology.orchestra.run.smithi155.stdout:            "id": 2,
 42506 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:            "name": "snap0"
 42507 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        },
 42508 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_synced": 0,
 42509 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_deleted": 0,
 42510 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_renamed": 0
 42511 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:    },
 42512 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:    "/d1": {
 42513 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "state": "idle",
 42514 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_synced": 0,
 42515 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_deleted": 0,
 42516 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_renamed": 0
 42517 2024-03-25T05:29:36.010 INFO:teuthology.orchestra.run.smithi155.stdout:    },
 42518 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:    "/d2": {
 42519 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        "state": "syncing",
 42520 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        "current_sycning_snap": {
 42521 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:            "id": 4,
 42522 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:            "name": "snap0"
 42523 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        },
 42524 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_synced": 0,
 42525 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_deleted": 0,
 42526 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:        "snaps_renamed": 0
 42527 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:    }
 42528 2024-03-25T05:29:36.011 INFO:teuthology.orchestra.run.smithi155.stdout:}
 42529 2024-03-25T05:29:36.017 DEBUG:tasks.cephfs.test_mirroring:command returned={
 42530     "/d0": {
 42531         "state": "syncing",
 42532         "current_sycning_snap": {
 42533             "id": 2,
 42534             "name": "snap0"
 42535         },
 42536         "snaps_synced": 0,
 42537         "snaps_deleted": 0,
 42538         "snaps_renamed": 0
 42539     },
 42540     "/d1": {
 42541         "state": "idle",
 42542         "snaps_synced": 0,
 42543         "snaps_deleted": 0,
 42544         "snaps_renamed": 0
 42545     },
 42546     "/d2": {
 42547         "state": "syncing",
 42548         "current_sycning_snap": {
 42549             "id": 4,
 42550             "name": "snap0"
 42551         },
 42552         "snaps_synced": 0,
 42553         "snaps_deleted": 0,
 42554         "snaps_renamed": 0
 42555     }
 42556 }
```
So the d1 didn't show `"state": "syncing"`, that's why it's failing. Btw, the newly pushed qa changes would wait in this case. 

Checking the ceph-client.mirror log shows `Bad file descriptor`:
```
 7498 2024-03-25T05:29:32.630+0000 7fbbdacb1640 -1 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) copy_to_remote: failed to write remote file path=./file.2: (9) Bad file descriptor
 7499 2024-03-25T05:29:32.630+0000 7fbbdacb1640 -1 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) copy_to_remote: failed to close remote fd path=./file.2: (9) Bad file descriptor
 7500 2024-03-25T05:29:32.630+0000 7fbbdacb1640 -1 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) remote_file_op: failed to copy path=./file.2: (22) Invalid argument
 7501 2024-03-25T05:29:32.630+0000 7fbbdacb1640 20 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) do_synchronize: closing local directory=.
 7502 2024-03-25T05:29:32.631+0000 7fbbdacb1640 20 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) post_sync_close_handles
 7503 2024-03-25T05:29:32.631+0000 7fbbdacb1640 -1 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) do_sync_snaps: failed to synchronize dir_root=/d1, snapshot=snap0
 7504 2024-03-25T05:29:32.631+0000 7fbbdacb1640 -1 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) sync_snaps: failed to sync snapshots for dir_root=/d1
 7505 2024-03-25T05:29:32.631+0000 7fbbdacb1640 20 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) unregister_directory: dir_root=/d1
 7506 2024-03-25T05:29:32.631+0000 7fbbdacb1640 20 cephfs::mirror::PeerReplayer(03077e74-e57d-47b6-be66-a4010927dcc5) unlock_directory: dir_root=/d1
```

Reply:
@joscollin In the failed run you reproduced, is the missing directory state (syncing) in status related to the bad file descriptor? It looks to me that it's related since the sync fails due to that or are you saying that the bad file descriptor is a different issue and this fix is just the delay of "syncing" status to show up in status?

Reply:
Sorry for my text caused a confusion.

As I checked in a passed test, the BADFD comes in the passed test also. It's within the synchronization task, but the BADFD hits at different parts of it.

See:
https://pulpito.ceph.com/jcollin-2024-03-22_09:24:18-fs:mirror-wip-jcollin-testing-22032024-distro-default-smithi/

Copy-pasting the parts:
```
 6268 2024-03-22T09:55:34.088+0000 7fbd9b036640  5 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) should_sync_entry: entry=./file.2, r=-2
 6269 2024-03-22T09:55:34.088+0000 7fbd9b036640  5 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: entry=./file.2, data_sync=1, attr_sync=1
 6270 2024-03-22T09:55:34.088+0000 7fbd9b036640 10 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) remote_file_op: dir_root=/d0, epath=./file.2, need_data_sync=1, need_attr_sync=1
 6271 2024-03-22T09:55:34.088+0000 7fbd9b036640 10 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) copy_to_remote: dir_root=/d0, epath=./file.2
 6272 2024-03-22T09:55:34.088+0000 7fbda8851640 20 client.6929 put_inode on 0x10000000000.2(faked_ino=0 nref=18 ll_ref=0 cap_refs={} open={0=1} mode=40755 size=0/0 nlink=1 btime=2024-03-22T09:53:50.109779+0000 mtime=2024-03-22T09:54:25.681267+0000 ctime=2024-03-22T09:55:32.820104+0000 change_attr=4 caps=pAsLsXsFs par\
      ents=0x10000000000.snapdir["snap0"] 0x559d137c2600) n = 1
 6273 2024-03-22T09:55:34.088+0000 7fbd9a835640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: 1 entries in stack
 6274 2024-03-22T09:55:34.088+0000 7fbd9a835640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: top of stack path=.
 6275 2024-03-22T09:55:34.088+0000 7fbd9a835640 10 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) propagate_deleted_entries: dir_root=/d2, epath=.
 6276 2024-03-22T09:55:34.088+0000 7fbda8851640 20 client.6929 put_inode on 0x10000000000.2(faked_ino=0 nref=18 ll_ref=0 cap_refs={} open={0=1} mode=40755 size=0/0 nlink=1 btime=2024-03-22T09:53:50.109779+0000 mtime=2024-03-22T09:54:25.681267+0000 ctime=2024-03-22T09:55:32.820104+0000 change_attr=4 caps=pAsLsXsFs par\
      ents=0x10000000000.snapdir["snap0"] 0x559d137c2600) n = 1
 6277 2024-03-22T09:55:34.088+0000 7fbd9a835640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: failed to propagate missing dirs: (9) Bad file descriptor
 6278 2024-03-22T09:55:34.088+0000 7fbd9a835640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: closing local directory=.
 6279 2024-03-22T09:55:34.088+0000 7fbd9a835640  3 client.6929 closedir(0x559d13417880) = 0
 6280 2024-03-22T09:55:34.088+0000 7fbd9a835640 10 client.6929 _closedir(0x559d13417880)
 6281 2024-03-22T09:55:34.088+0000 7fbd9a835640 10 client.6929 _closedir detaching inode 0x559d137c2000
================================

 18429 2024-03-22T09:57:03.902+0000 7fbd9b036640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_sync_snaps: failed to synchronize dir_root=/d1, snapshot=snap0
 18430 2024-03-22T09:57:03.902+0000 7fbd9b036640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) sync_snaps: failed to sync snapshots for dir_root=/d1
 18431 2024-03-22T09:57:03.902+0000 7fbd9b036640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) unregister_directory: dir_root=/d1
 18432 2024-03-22T09:57:03.902+0000 7fbd9b036640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) unlock_directory: dir_root=/d1
 18433 2024-03-22T09:57:03.902+0000 7fbd9b036640 10 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) unlock_directory: dir_root=/d1 unlocked
 18434 2024-03-22T09:57:03.903+0000 7fbd9b837640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) copy_to_remote: failed to read local file path=./file_37: (9) Bad file descriptor
 18435 2024-03-22T09:57:03.903+0000 7fbd9b837640  3 client.6929 close enter(10)
 18436 2024-03-22T09:57:03.903+0000 7fbd9b837640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) copy_to_remote: failed to close local fd path=./file_37: (9) Bad file descriptor
 18437 2024-03-22T09:57:03.903+0000 7fbd9b837640 -1 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) remote_file_op: failed to copy path=./file_37: (22) Invalid argument
 18438 2024-03-22T09:57:03.903+0000 7fbd9b837640 20 cephfs::mirror::PeerReplayer(5baea330-232d-4f14-9c5b-73505355e334) do_synchronize: closing local directory=.
```
So it would be unrelated to the issue this PR fixes. Let me check a few more failed jobs and create a new tracker for it.

Reply:
@vshankar Created https://tracker.ceph.com/issues/65115.